[c10d] Move the timeout dump check from watchdog to monitoring thread

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117168

To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

fduwjj committed:
[c10d] Move the timeout dump check from watchdog to monitoring thread

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check to monitoring thread. We also need to ensure that 

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1265,20 +1265,56 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
       return;
     }
 
+    // Assuming that we always init a process group containing all ranks,
```
"We put extra functionality in the thread for the default PG (uid_=0) any time we need something to run once per process and avoid doing it from too many separate threads.  For example, we check a global flag on the TCPStore to see if anyone else observed timeout and signaled peers to dump state, and we avoid hammering the TCPStore from all threads"  might be clearer

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1265,20 +1265,56 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              (currentTime - lastTimePollStore))
+              .count();
+      if (timeSinceLastWorkListUpdate >= kWatchdogThreadSleepMillis &&
```
this whole block is so hard to read.  i am trying to think of a way to make it briefer.

what if we made a helper for computeDeltaMS(start, end) = std::chrono::duration_cast<std::chrono::milliseconds>((end - start)).count()

and then we do 

if(computeDeltaMS(lastWorkListUpdateTime_ , currentTime) >= kWatchdogThreadSleepMillis &&
    computeDeltaMS(...) >=  ...) {

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3995,7 +3996,7 @@ def test_timeout_dumps_on_idle_ranks(self):
             # Force rank 1 to idle so that it also gets debug info dump triggered.
             if self.rank == 1:
-                time.sleep(6)
+                time.sleep(8)
```
I see.. Hmm I think that is a better idea haha. We already have the set timeout API ready, let me do that.

fduwjj replied:
_set_pg_timeout will change the timeout for both... Hmm I probably just make it sleep super long.

fduwjj replied:
Actually if I let rank 1 to call a different collective, now watchdog timeout for rank 0 does not file.. (The process now returns 0), which I cannot explain..

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1239,22 +1239,64 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+
 void ProcessGroupNCCL::heartbeatMonitor() {
   uint64_t heartBeatCounter = 0ULL;
+  int heartBeatTimeout = (dumpOnTimeout_ && uid_ == 0)
```
Exactly... I thought about the idea you mentioned here as well. I originally didn't want to make the code look messy.. Let me see if we can do it cleanly.

wconstab replied:
yea, it will at least be cleaner now that you have the helper methods, but bottom line we shouldn't introduce accidental behavior changes

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3977,10 +3977,10 @@ def test_timeout_dumps_on_idle_ranks(self):
-        os.environ[
-            "TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC"
-        ] = f"{NCCLTraceTestDumpOnTimeoutBase.timeout_sec * 2}"
+        # # Set heartbeat timeout to a shorter one (default timeout is 10 mins).
```
intended to keep this commented out? or want to add it back?

fduwjj replied:
Hmm this should be deleted in the latest version.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1248,10 +1248,11 @@ int computeDeltaMS(
-      ? timeoutCheckIntervalMilSec_
-      : heartbeatTimeoutInSec_ * 1000;
+  bool checkTimeoutSignal = (dumpOnTimeout_ && uid_ == 0);
+  int heartBeatTimeout = checkTimeoutSignal ? timeoutCheckIntervalMilSec_
```
maybe rename `heartBeatTimeout` to `monitorPollInterval`?

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1239,29 +1239,77 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
-    } else {
-      // No heartbeat increase detected and timeout.
-      break;
+    if (!checkTimeoutSignal ||
```
imo clearer to always compute time rather than have the || condition (which is correct but looks like an optimization we might not need)

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1239,29 +1239,77 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+        lastTimePollStore = currentTime;
+        if (globalStore_->check({std::string(TIMEOUT_DUMP)})) {
+          auto wakeUpTime = getWakeupTime(waitTimeoutDumpInMilSec_);
+          asyncDebugDump = launchAsyncDebugDump();
```
do we need to have the code for launching/dumping inside this if?

why not just 'break' here and rely on the normal exit path below to do the dump?

We can add one log here saying "received dump signal" so we can tell that that's why we dumped rather than actual heartbeat failure.

We can also add a log in the heartbeat comparison saying "heartbeat missed"

or we can create an empty string at the top, set it to either "dump signaled by other rank", or "heartbeat not detected"

then in the codepath below we can just append this "reason" log to the main log we emit

fduwjj replied:
Hmmm, but that would lead to program aborted instead of exception throwing. We have no exception throwing anyway. Maybe this should be fine?

wconstab replied:
if you want to throw in one case and abort in other case, can you say why?  (or should we just be consistent?)

fduwjj replied:
Well.. you are right, this exception is a legacy from watchdog where we only throw exception while in monitoring thread, we just do abort. But I don't know if switching to abort will make it testable or not. Let me try.

fduwjj replied:
Oh Abort actually works for unit test, nice! OK I think I can make this change.

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3924,6 +3924,7 @@ class NCCLTraceTestDumpOnTimeout(NCCLTraceTestDumpOnTimeoutBase):
     @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
     @parametrize("timing_enabled", [True, False])
     def test_timeout_dumps(self, timing_enabled):
+        os.environ['TORCH_NCCL_COORD_DUMP_CHECK_MS'] = '10000'
```
can you comment why this setting?  (It's not obvious if you are choosing a long time (10sec seems long-ish) so that coordination does _not_ happen during this test, or if you are choosing a short time so that it happens during the test (and 10sec doesn't seem very short)   But in general, how do you convince me this will be reliable and not noisy test?

fduwjj replied:
Sure. Let me add a comment here. Basically, I found that if we check too frequent, rank1 will also timeout so that it does not return 0 for exit code.

fduwjj replied:
This test does not test the coordinated timeout dump or it does not care. The unit test tests the feature of local timeout dump and it assume another rank will exit smoothly. But if we check too frequently, another rank will also timeout.

wconstab reviewed (approved):
i think this looks good. @zdevito do you want to take a look too?

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3924,6 +3924,8 @@ class NCCLTraceTestDumpOnTimeout(NCCLTraceTestDumpOnTimeoutBase):
     @parametrize("timing_enabled", [True, False])
     def test_timeout_dumps(self, timing_enabled):
+        # We need to reduce the coordinated check frequency to avoid rank 0 also timeout.
+        os.environ['TORCH_NCCL_COORD_CHECK_MILSEC'] = '15000'
```
ok. Given that these tests are generally pretty slow (many tests take 10-20 second to run), choosing a number like '15s' for "a long enough time to not happen" looks risky - thats what originally motivated me to ask this question.

IF you are just trying to disable coordination, then i'd suggest setting a REALLY large number and update the comment to be explicit that you're trying to avoid coordinated dump so that only rank0 reports timeout

fduwjj replied:
Sure. Let me do that.

fduwjj committed:
Update on "[c10d] Move the timeout dump check from watchdog to monitoring thread"


To avoid potential hang in watchdog thread which will prevent us from dumping timeout debugging info, we move the check of global collective timeout signals and dumping debugging info to monitoring thread. We also need to ensure that we don't wait very long to check out the timeout signal from store; otherwise, we will miss the signal and don't get debugging info dumped.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]
